web/src/app/chat/components/input/ActionManagement.tsx

```diff
@@ -439,79 +481,255 @@ function MCPToolsList({
-              text-text-700
-              placeholder:text-text-400
-              dark:placeholder:text-neutral-600
-              dark:bg-neutral-950
-            "
-            autoFocus
-          />
+    <>
+      <div className="bg-transparent flex-shrink-0">
+        <div className="mx-1">
+          <div className="relative">
+            <button
+              onClick={() => {
+                setSearchTerm("");
+                onBack();
+              }}
+              className="absolute left-1 top-1/2 transform -translate-y-1/2 text-text-400 hover:text-text-300 z-10 w-4 h-4 flex items-center justify-center transition-colors"
+              style={{ borderRadius: "8px" }}
+              aria-label="Back to action menu"
+            >
+              <FiChevronLeft size={16} />
+            </button>
+            <input
+              type="text"
+              placeholder={`Search ${placeholderName} tools`}
+              value={searchTerm}
+              onChange={(e) => setSearchTerm(e.target.value)}
+              className="
+                w-full
+                pl-7
+                pr-3
+                py-2
+                bg-transparent
+                rounded-lg
+                text-sm
+                outline-none
+                text-neutral-700 dark:text-neutral-300
+                placeholder:text-neutral-400 dark:placeholder:text-neutral-500
+              "
+              autoFocus
+            />
+          </div>
         </div>
+        <div className="mx-1">
+          <div className="relative mt-1">
+            <input
+              type="text"
+              placeholder={
+                allToolsDisabled ? "Enable All Tools" : "Disable All Tools"
+              }
+              readOnly
+              onClick={
+                allToolsDisabled
+                  ? enableAllToolsForServer
+                  : disableAllToolsForServer
+              }
+              className="
+                w-full
+                pl-7
+                pr-9
+                py-2
+                bg-transparent
+                rounded-lg
+                text-sm
+                outline-none
+                text-neutral-700 dark:text-neutral-300
+                placeholder:text-neutral-400 dark:placeholder:text-neutral-500
+                cursor-pointer
+              "
+            />
+            <img
+              src={allToolsDisabled ? "/plug.svg" : "/unplug.svg"}
+              alt={allToolsDisabled ? "Enable All Tools" : "Disable All Tools"}
+              className="absolute right-3 top-1/2 transform -translate-y-1/2 w-4 h-4 pointer-events-none"
+              style={{
+                filter:
+                  theme === "dark"
+                    ? "invert(0.6) sepia(1) saturate(0) hue-rotate(0deg) brightness(1.2)"
+                    : "invert(0.6) sepia(1) saturate(0) hue-rotate(0deg) brightness(0.6)",
+                opacity: theme === "dark" ? 0.9 : 0.8,
+              }}
+            />
```

Comment:
Using `img` to render SVGs is usually bad form.

@duo-onyx has created the appropriate plug and unplug SVGs in Figma. You can import them into `web/src/icons` by running through the manual instructions laid out in `web/src/icons/README.md`. OR, you could instead ask the Figma MCP tool to directly import them into the `web/src/icons` directory!

Note that if you opt for the latter, you will need to either tell the AI tool to remove the `stroke="..."` and `strokeOpacity={...}` props, or do it yourself after import. Those props hardcode the stroke colour and opacity, thus making it impossible for the end-user to specify it themselves.